FIX `SSViwer::execute_template` should allow requirements to be included

dhensby opened the pull request:
Potential fix to #7729 

This creates a `Requirements_Backend` instance for the rendering of template when using `SSViewer::execute_template` so that requirements can be safely included in templates that are rendered standalone.

tractorcow commented:
If you are using `<% require SomeTemplatesWithIncludes %>` I would (as a developer) prefer that these requirements were added to the top level template, which may actually contain my `<head>` tag (rather than forcing these inline).

E.g.

```
<html>
<head>
</head>
<body>
<% Include SomeTemplateWithIncludes %>
</body>
</html>
```

Should probably allow those css / js required files to go into the `<head>`. See my comments on https://github.com/silverstripe/silverstripe-framework/issues/7729 about how I see those `<% require %>` calls best behaving.

At the moment, those files aren't added anywhere, so it's still broken as it is. :)

Might I suggest instead another flag added to execute_template() to toggle the behaviour of require.

```
/**
     * Execute the given template, passing it the given data.
     * Used by the <% include %> template tag to process templates.
     *
     * @param string $template Template name
     * @param mixed $data Data context
     * @param array $arguments Additional arguments
     * @param Object $scope
     * @param bool $inlineRequirements Include required resources in the result. If false, these will be added to the global Requirements_Backend object instead
     * @return string Evaluated result
     */
    public static function execute_template($template, $data, $arguments = null, $scope = null, $inlineRequirements = true)
```

Note that the default matches your suggested behaviour; Most of the times that `execute_template` is called by users, those users would expect those files being included.

however, SSTemplateParser would pass in `false` to this argument, meaning that all requirements for a single template call (containing many `<% includes %>` potentially) are collected into a single top level requirements.


```php
// Note: 'type' here is important to disable subTemplates in SSViewer::getSubtemplateFor()
// Pass in false to $inlineRequirements to ensure requirements are collected at a single top level
        $res['php'] = '$val .= \\SilverStripe\\View\\SSViewer::execute_template([["type" => "Includes", '.$template.'], '.$template.'], $scope->getItem(), array(' .
            implode(',', $arguments)."), \$scope, false);\n";
```

And thirdly, get rid of `$v->includeRequirements(false);` call in execute_template() completely (as you have done), since a `<% require %>` should never be ignored.

tractorcow commented:
Note that the other risk is that multiple templates may `<% require %>` the same file (e.g. jquery) so consolidating this into a single requirements captures that use case.

dhensby commented:
ah, I didn't realise that `execute_template` was used so extensively in the compiled templates!

The issue I'm wanting to resolve is the fact that when used as a standalone template executor, the requirements are not included and this is a larger problem now that module resource location is managed by using requirements instead of being hard coded into templates...

I'll look into your suggested solution @tractorcow, thanks :)

tractorcow commented:
I wonder if $scope should actually have a requirements object attached to it, and if not passed it it adds it to global requirements.

dhensby commented:
OK, I've added an extra parameter to the `execute_template` and `execute_string` methods which toggles this behaviour and I've updated the template parser to pass through the right argument.

dhensby commented:
Sure

dhensby commented:
rebased

tractorcow commented:
Um, you rebased and deleted all the new code @dhensby .

> OK, I've added an extra parameter to the execute_template and execute_string methods which toggles this behaviour and I've updated the template parser to pass through the right argument.

That code seems to have disappeared. =(

dhensby commented:
Well that wasn't clever... I'll sort it out

dhensby commented:
OK - back to how it should be :)

tractorcow commented:
Test failures unrelated, fixed with https://github.com/silverstripe/silverstripe-framework/pull/7824
